src/chains/ethereum/ethereum/tests/api/eth/logs.test.ts

```diff
@@ -62,39 +75,145 @@ describe("api", () => {
             await provider.send("eth_sendTransaction", [{ ...tx }]);
-            const message = await Promise.race([
-              onceMessageFor(subscriptionId),
-              onceMessageFor(subscriptionId2)
+            const messages = await Promise.race([
+              getMessagesForSub(subscriptionId, numberOfLogs, []),
+              getMessagesForSub(subscriptionId2, numberOfLogs, [])
+            ]);
+            // the one to return for all messages is the sub2, which we never unsubed
+            for (let i = 0; i < numberOfLogs; i++) {
+              assert.strictEqual(
+                messages[i].data.subscription,
+                subscriptionId2,
+                "unsubscribe didn't work"
+              );
+            }
+          });
+
+          it("filters subscription by address", async () => {
+            const subscriptionId = await provider.send("eth_subscribe", [
+              "logs",
+              { address: contractAddress }
+            ]);
+
+            assert(subscriptionId != null);
+            assert.notStrictEqual(subscriptionId, false);
+
+            const secondContractAddress = await deployContractAndGetAddress();
+
+            // trigger a log event, we should get four events
+            const numberOfLogs = 4;
+            const data =
+              "0x" +
+              contract.contract.evm.methodIdentifiers["logNTimes(uint8)"] +
+              numberOfLogs.toString().padStart(64, "0");
+            // tx to accounts other than our filter address should not appear in logs
+            const filteredTx = {
+              from: accounts[0],
+              to: secondContractAddress,
+              data
+            };
+            const loggedTx = { from: accounts[0], to: contractAddress, data };
+
+            // ensure subscription is working and we can receive logs sent to the original contract
+            const logged = getMessagesForSub(subscriptionId, numberOfLogs, []);
+            await provider.send("eth_sendTransaction", [{ ...loggedTx }]);
+
+            const messages = await logged;
+            assert.strictEqual(messages.length, numberOfLogs);
+            for (let i = 0; i < numberOfLogs; i++) {
+              assert.strictEqual(
+                messages[i].data.result.address.toString(),
+                contractAddress,
+                "log subscription filtering by address didn't return correct results"
+              );
+            }
+            // ensure filtering is working and we don't receive logs from txs sent somewhere other than what
+            // we put in our filter
+            const filtered = getMessagesForSub(
+              subscriptionId,
+              numberOfLogs,
+              []
+            );
+            await provider.send("eth_sendTransaction", [{ ...filteredTx }]);
+
+            const noMessages = await Promise.race([
+              filtered,
+              new Promise(resolve => setTimeout(resolve, 3000, "timeout"))
```

Comment:
Interesting approach here. I wonder if there is a better way? Maybe listening for a `newHeads` event would work? Though i'm not sure that would guarantee what we are trying to guarantee here. hmmmmmm :thinking:

Reply:
It looks like `.emit("blockLogs", blockLogs)` is always emitted immediately before `.emit("block", block)`, so presumably `logs` subscriptions would beat `newHeads` in a race, unless they were filtered. Probably is a bit frail though.

Reply:
Yeah, it sounds like we'd also need to write a test here that ensures the order of messages here :-/ hmmm

Reply:
See updated code, we've managed to work around the timeout.